Add judge overview (#231)
dodona-edu/dodona-edu.github.io: en/references/judges/README.md

---
title: Judges
description: "Overview of all judges available on Dodona"
---

# Judges

This guide is written for teachers who are creating exercises for Dodona.

Before you create an exercise, you should decide which `judge` you want to use.
The `judge` is the program that will evaluate the submissions of students.

> It is often written for one specific programming language, or for a set of programming languages.

Reviewer: ```suggestion
It is often written for one specific programming language, but supporting a set of programming languages is also possible.
```

> Each `judge` has its own configuration options, we will provide a link to the relevant documentation for each `judge` below.

Reviewer: ```suggestion
Each judge has its own configuration options, we will provide a link to the relevant documentation for each judge below.
```

> These judge specific options should be provided in the `evaluation` directory of the [exercise configuration](/en/references/exercise-directory-structure).

Reviewer: ```suggestion
These judge-specific options should be provided in the `evaluation` directory of the [exercise configuration](/en/references/exercise-directory-structure).
```

Advanced users can also create their own judge, see [this guide](/en/guides/creating-a-judge).

Dodona currently supports the following judges:

## TESTed

> TESTED is a whitebox judge that can be used for multiple programming languages.

Reviewer: ```suggestion
TESTed is a whitebox judge that can be used for multiple programming languages.
```

> It uses a simple custom test format, that is indpendent of the programming language.\

Reviewer: ```suggestion
It uses a simple custom test format, that is independent of the programming language of the exercise.\
```

> **Programing languages:** Bash, C, C#, Haskell, Java, Javascript, Kotlin, Python\

Reviewer: Do we want to do this like this, or should we look into a definition list, e.g. https://github.com/markdown-it/markdown-it-deflist?

Author: As discussed I will ignore this one, because it does not really improve the visuals, and makes each judge take up a lot more space
![image](https://user-images.githubusercontent.com/21177904/218117641-2c09d22d-fb79-4f65-9fcc-46226a64b3e9.png)

> **Programing languages:** Bash, C, C#, Haskell, Java, Javascript, Kotlin, Python\

Reviewer: ```suggestion
**Programming languages:** Bash, C, C#, Haskell, Java, JavaScript, Kotlin, Python\
```

**Get started** [Documentation](/en/tested#designing-exercises-for-dodona), [examples](https://github.com/dodona-edu/universal-judge/tree/master/exercise) \

> **Created by:** [Team dodona](mailto:dodona@ugent.be)

Reviewer: ```suggestion
**Created by:** [Team Dodona](mailto:dodona@ugent.be)
```

This should be done everywhere in this file.

## Python
Python/Pythia is the first judge that was created for Dodona.

> It is a python judge that allows simple input/output tests or more advanced doctests.\

Reviewer 2: ```suggestion
It is a Python judge that allows simple input/output tests or more advanced doctests.\
```

> **Programing languages:** Python\

Reviewer: ```suggestion
**Programming languages:** Python\
```

This should also be done everywhere.

**Get started** [Documentation](/en/references/python-judge), [examples](https://github.com/dodona-edu/example-exercises/tree/master/python) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## R
R is a judge that can be used for exercises on the R programming language.\
**Programing languages:** R\
**Get started** [Documentation](https://github.com/dodona-edu/judge-r), [examples](https://github.com/dodona-edu/example-exercises/tree/master/R) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## Java

> The java judge uses the JUnit4 framework to run tests on java exercises.\

Reviewer 3: ```suggestion
The java judge uses the JUnit4 framework to run tests on Java exercises.\
```

**Programing languages:** C\
**Get started** [Documentation](https://github.com/mvdcamme/C-Judge), [examples](https://github.com/mvdcamme/C-Judge/tree/master/example_exercises) \
**Created by:** [Maarten Vandercammen](mailto:mvdcamme@vub.ac.be)

## SQL
The SQL judge supports both query evaluation (DML) and structural database building (DDL).\
**Programing languages:** SQL\
**Get started** [Documentation](https://github.com/dodona-edu/judge-sql), [examples](https://github.com/dodona-edu/example-exercises/tree/master/sql) \
**Created by:** [Brecht Willems](mailto:Brecht.Willems@UGent.be)

## Scheme
Scheme is a judge that supports the `R5RS` variant of the scheme programming language. It uses a custom testframework [dunit](https://gitlab.soft.vub.ac.be/Structuur1/dodona-judge/-/tree/master/collects/dunit) to define the tests.\
**Programing languages:** Scheme\
**Get started** [Documentation](https://gitlab.soft.vub.ac.be/Structuur1/dodona-judge), [examples](https://gitlab.soft.vub.ac.be/Structuur1/dodona-judge/-/tree/master/example-exercises) \
**Created by:** [Mathijs Saey](mailto:scpi@dinf.vub.ac.be)

## Prolog

> Prolog is a judge that can be used for exercises on the prolog programming language.

Reviewer: ```suggestion
Prolog is a judge that can be used for exercises on the Prolog programming language.
```

It supports PLUnit, QuickCheck and simple input output tests.\
**Programing languages:** Prolog\
**Get started** [Documentation](https://github.com/dodona-edu/judge-prolog), [examples](https://github.com/dodona-edu/example-exercises/tree/master/prolog) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## Haskell

> Haskell is a judge that uses HUnit to test haskell exercises. \

Reviewer 3: ```suggestion
Haskell is a judge that uses HUnit to test Haskell exercises. \
```

**Programing languages:** Scheme\
**Get started** Contact the creators to get more info about this judge.\
**Created by:** [Mathijs Saey](mailto:mathijs.saey@vub.be)

## HTML
The HTML judge evaluates both the HTML and CSS code of a student, based on a model solution or a checklist of criteria.\
**Programing languages:** HTML, CSS\
**Get started** [Documentation](https://github.com/dodona-edu/judge-html), [examples](https://github.com/dodona-edu/example-exercises/tree/master/html) \
**Created by:** [Brecht Willems](mailto:Brecht.Willems@UGent.be)

## Turtle

> The Turtle judge evaluates the output of a python turtle program. It calculates the similarity between the output of the student and the model solution.\

Reviewer 2: ```suggestion
The Turtle judge evaluates the output of a Python turtle program. It calculates the similarity between the output of the student and the model solution.\
```

> The Turtle judge evaluates the output of a Python turtle program. It calculates the similarity between the output of the student and the model solution.\

Reviewer: ```suggestion
The Turtle judge evaluates the output of a Python Turtle program. It calculates the similarity between the output of the student and the model solution.\
```

**Programing languages:** Python (Turtle)\
**Get started** [Documentation](https://github.com/dodona-edu/judge-turtle) \
**Created by:** [Brecht Willems](mailto:Brecht.Willems@UGent.be)

## Markdown
The markdown judge is not a real judge as it does not evaluate code.

> It does render the markdown code of a student and can be useful to manually evaluate the output in Dodona.\

Reviewer: ```suggestion
It does render the Markdown code of a student and can be useful to manually evaluate the output in Dodona.\
```

**Programing languages:** Markdown\
**Get started** [Documentation](https://github.com/dodona-edu/judge-markdown) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## Javascript

> Javascript is a judge that can be used for exercises on the javascript programming language.

Reviewer 3: ```suggestion
Javascript is a judge that can be used for exercises on the JavaScript programming language.
```

It is undocumented and has a lot of very usecase specific implementations.

> If you want to create your own javascript exercises, we recommend you to use the [TESTed judge](#tested) instead.\

Reviewer: ```suggestion
If you want to create your own JavaScript exercises, we recommend you to use the [TESTed judge](#tested) instead.\
```

> **Programing languages:** Javascript\

Reviewer: ```suggestion
**Programming languages:** JavaScript\
```

**Get started** [Github repo](https://github.com/dodona-edu/judge-javascript), [examples](https://github.com/dodona-edu/example-exercises/tree/master/javascript) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## Bash
Bash is a judge that can be used for exercises on the bash terminal.
It is undocumented and has a lot of very usecase specific implementations.

> If you want to create your own bash exercises, we recommend you to use the [TESTed judge](#tested) instead.

Reviewer: ```suggestion
If you want to create your own Bash exercises, we recommend you to use the [TESTed judge](#tested) instead.
```

**Programing languages:** Bash\
**Get started** [Examples](https://github.com/dodona-edu/example-exercises/tree/master/bash), contact the creators to get more info about this judge. \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## Csharp (Deprecated)
The Csharp judge is deprecated and should only be used for legacy exercises.
If you want to create your own C# exercises, we recommend you to use the [TESTed judge](#tested) instead.\
**Programing languages:** C#\
**Get started** [Examples](https://github.com/dodona-edu/example-exercises/tree/master/c%23) \
**Created by:** [Team dodona](mailto:dodona@ugent.be)

## JUnit (Deprecated)

> The JUnit judge is a judge for java8 exercises.

Reviewer 3: ```suggestion
The JUnit judge is a judge for Java 8 exercises.
```
